src/olympia/files/models.py

```diff
@@ -378,40 +378,53 @@ def get_localepicker(self):
     def webext_permissions(self):
         """Return permissions that should be displayed, with descriptions, in
         defined order:
-        1) match all urls (e.g. <all-urls>)
-        2) known permissions, in constants order (alphabetically),
-        3) match urls for sites
+        1) Either the match all permission, if present (e.g. <all-urls>), or
+           match urls for sites (<all-urls> takes preference over match urls)
+        2) nativeMessaging permission, if present
+        3) other known permissions in alphabetical order
         """
         knowns = list(WebextPermissionDescription.objects.filter(
             name__in=self.webext_permissions_list).iterator())
-
-        match_all, urls = [], []
+        # Move nativeMessaging permission to front of the list, if present
+        for index, perm in enumerate(knowns):
+            if perm.name == WebextPermission.NATIVE_MESSAGING_NAME:
+                perm.description.localized_string = (
+                    perm.description.localized_string.replace('%S',
+                                                              _('Firefox')))
+                knowns.pop(index)
+                knowns.insert(0, perm)
+                break
+
+        urls = []
         for name in self.webext_permissions_list:
             if re.match(WebextPermissionDescription.MATCH_ALL_REGEX, name):
-                # Add match-alls.  Yes, assign not append - no dupes plz.
-                match_all = [WebextPermissionDescription.ALL_URLS_PERMISSION]
+                # We found a match-all, so no need to continue.
+                return (
+                    [WebextPermissionDescription.ALL_URLS_PERMISSION] + knowns)
```

Comment:
I think the early return is now really getting in the way of readability. The method is now quite big, to scanning for early returns is harder...